### PR TITLE
treewide: wallpaper refactor

### PR DESCRIPTION
**Pull request by SomeGuyNamedMay**
# Checklist

- [x] Module support
- [x] Check function for types that actually restricts the type
- [x] Add documentation
- [ ] Add testing

# Potential Additions

- Move more configuration into the custom types such as polarity
- Additional constructors for things such as generating a wallpapers like penrose tiling's and the like
- Add a utility function for documenting constructors

# Reference

The following depend or should be improved based on this PR:

- https://github.com/danth/stylix/issues/131
- https://github.com/danth/stylix/issues/182
- https://github.com/danth/stylix/issues/207
- https://github.com/danth/stylix/issues/208
- https://github.com/danth/stylix/issues/251
- https://github.com/danth/stylix/issues/63
- https://github.com/danth/stylix/pull/135
- https://github.com/danth/stylix/pull/157
- https://github.com/danth/stylix/pull/175

### CODE REVIEW

File: stylix/palette.nix
Comment (danth) [inline stylix/palette.nix:27]: I think it's possible to have this transfer the image to the new option, and print a warning, rather than an error. That would avoid breaking old configs.

File: stylix/palette.nix
Comment (danth) [inline stylix/palette.nix:47]: Probably neater to use a `with` here

```suggestion
        with types; with config.lib.stylix; oneOf [static animation video slideshow];
```

**Comment by SomeGuyNamedMy** [comment]
running into this weird error when building docs
[error.txt](https://github.com/danth/stylix/files/11682049/error.txt)

**Comment by SomeGuyNamedMy** [comment]
ready for review when you've got time.

**Comment by SomeGuyNamedMy** [comment]
i believe any window manager that worked with feh should work with the animations?

**Comment by SomeGuyNamedMy** [comment]
@danth i think this is ready whenever you have time

File: lib/types.nix
Comment (danth) [inline lib/types.nix:16]:
```suggestion
      descriptionClass = "animated image such as a GIF";
```

File: lib/types.nix
Comment (danth) [inline lib/types.nix:40]: Might be neater to put these with the corresponding type rather than the bottom of the file.

File: lib/utils.nix
Comment (danth) [inline lib/utils.nix:1]: I'd recommend writing these inside a `config.lib.stylix = { }` bracket rather than repeating `config.lib.stylix` on every line

File: modules/sway/hm.nix
Comment (danth) [inline modules/sway/hm.nix:29]: Typo
```suggestion
              $0 <dir containing images>"
```

File: modules/sway/hm.nix
Comment (danth) [inline modules/sway/hm.nix:26]: This slideshow script is repeated in a few places, I've only reviewed this copy, maybe we should move it to a central file if possible

File: modules/sway/hm.nix
Comment (SomeGuyNamedMy) [reply modules/sway/hm.nix]: i ended up using a different solution to simplify the script

File: modules/sway/hm.nix
Comment (SomeGuyNamedMy) [reply modules/sway/hm.nix]: i think moving the wayland one to lib would be a good idea cause that can be used in other compositors like river, however the xorg version i can only really see working through xsession.

**Comment by danth** [comment]
We need to update the configuration page on the website to use the new constructors

**Comment by SomeGuyNamedMy** [comment]
do we want to run checks on that wallpaper types that ensures there contents are of the correct types?

**Comment by danth** [comment]
That would be good if possible

**Comment by danth** [comment]
I'm wondering if we should rename the option from `wallpaper` to something like `theme` or `scheme`, since it also contains the colour scheme and some other settings not related to the wallpaper.

**Comment by SomeGuyNamedMy** [comment]
yea i thought about that as well, ill differ to you on that one.

**Comment by danth** [comment]
I found a small problem if you set a wallpaper on Wayland which is wider than the screen.

(I'm running Hyprland so it's not directly using your code, but I copied the `wbg` command into my config.)

wbg wallpaper

It would look better if the image was cropped rather than squashed, which is what `swww` does by default.

swww wallpaper

We already use `swww` for slideshows and animations, so perhaps it would make sense to use it for static images too - then we also get a smooth transition if the config is changed.

`swaybg` also supports the crop if you run it with `swaybg --image «path to image» --mode fill`.

**Comment by SomeGuyNamedMy** [comment]
@danth do you mind if i add hyprland support to this pr as well? it would make to wlroots side of things easier for me to test

**Comment by danth** [comment]
I feel like Hyprland might be better left for a separate pull request - this one is almost ready to merge.

**Comment by danth** [comment]
Hi! Sorry for the radio silence, I just merged Hyprland support using the new Home Manager module if you were wanting to use that for the wallpapers

**Comment by SomeGuyNamedMy** [comment]
i went and looked on the kde side and cant find a way do set animations using nix, they appear to be available using plugins though, maybe we could place the image/animation/video in .config/stylix for desktops that we cant directly support?

**Comment by SomeGuyNamedMy** [comment]
also i went looking into adding the extra utility functions to lib, it appears to be possible, but difficult to implement, since we use them both as a user facing interface, as well as internally.

**Comment by danth** [comment]
Although we can't add things to `lib` we should be able to make a separate `stylix` input so you can write `stylix.function` rather than `config.lib.stylix.function`. I was planning to commit that as soon as I have time to work on it, possibly later today

**Comment by danth** [comment]
It looks like anywhere we use `swww`, if you set a slideshow wallpaper then change it to a different slideshow, the script will end up running twice until you restart. Could we find a way to fix this? Maybe moving it to a systemd user service would help.

**Comment by SomeGuyNamedMy** [comment]
i went and checked on the kde side of things, it appears that we would somehow have to install a plugin to use a non static wallpaper in kde, i dont have much experience with the kde nix inferstructure, so im not sure how, or if you even can, install kde plugins using nix.

**Comment by danth** [comment]
In my experience with KDE on NixOS there's not much you can set up declaratively beyond basic settings. There might be a way to install the plugin by running commands from an activation script, user service or whatever, but that's not so reliable.

**Comment by SomeGuyNamedMy** [comment]
went ahead and put the wallpaper in etc/stylix and .config/stylix for unsupported modules

File: modules/wlroots/hm.nix
Comment (trueNAHO) [inline modules/wlroots/hm.nix:29]: Is `PATH` required in `swww.Service.Environment` when `swww.Service.ExecStart` and `swww.Service.ExecStop` already use the absolute `/nix/store/<swww>/bin/sww` path?

**Review by trueNAHO — CHANGES_REQUESTED** [review]

File: modules/xorg/nixos.nix
Comment (trueNAHO) [inline modules/xorg/nixos.nix:30]: Should we not refactor this logic into an internal module because it is very similar to what is done in `modules/xorg/hm.nix`.

File: modules/xorg/hm.nix
Comment (trueNAHO) [inline modules/xorg/hm.nix:31]: Should we refactor into a variable:

```suggestion
 config.xsession.initExtra = let 
   command = value: ''
     ${pkgs.xwinwrap}/bin/xwinwrap -fs -ni -b -nf -ov -- mpv -wid WID --loop --no-audio ${value}
   ''; 
 in lib.mkIf config.stylix.targets.xorg.enable 
   (if (isAnimation config.stylix.wallpaper) then (command config.stylix.wallpaper.animation) else
     (if (isVideo config.stylix.wallpaper) then (command config.stylix.wallpaper.video) else
```

File: docs/settings.nix
Comment (trueNAHO) [inline docs/settings.nix:1]:
```suggestion
# Dummy wallpapers to avoid errors when generating the documentation.N
```

**Review by trueNAHO — CHANGES_REQUESTED** [review]

File: docs/settings.nix
Comment (danth) [reply docs/settings.nix:1]: There is only one wallpaper defined here so it does not need to be plural.

File: lib/constructors.nix
Comment (danth) [reply lib/constructors.nix:15]: Could this include the `pkgs.runCommand` part too?

File: lib/types.nix
Comment (danth) [reply lib/types.nix:5]: This may have a small effect, but I think it would cause too much confusion to be worth doing.

File: modules/rofi/hm.nix
Comment (danth) [reply modules/rofi/hm.nix:10]: `with config.lib.stylix;` may be more appropriate

File: modules/wlroots/hm.nix
Comment (danth) [reply modules/wlroots/hm.nix:13]: That sounds like a very good idea. However, I'm not sure why images are being selected at random here. Would users expect the slideshow to follow the order they choose?

Comment (trueNAHO) [reply modules/wlroots/hm.nix:13]: > However, I'm not sure why images are being selected at random here. Would users expect the slideshow to follow the order they choose?

Sounds like this should be a Stylix option.

> Using a timer will only work with randomness, as otherwise we would have no way to remember the index of the current image.

We could cache statefull information in files, like `/tmp/` or `/run`: https://superuser.com/questions/1341811/stop-a-systemd-service-and-clear-all-persistent-data

Comment (danth) [reply modules/wlroots/hm.nix:13]: That seems like a lot of extra complication compared to a simple `while` loop.

File: modules/wlroots/hm.nix
Comment (danth) [reply modules/wlroots/hm.nix:29]: I'm not sure whether that syntax is understood by systemd - `ExecStart` isn't read like a normal shell command.

File: modules/wlroots/hm.nix
Comment (trueNAHO) [reply modules/wlroots/hm.nix:13]: Sadyly, the `while` loop has the same disadvantages as cron jobs compared to Systemd timers.

Particularly, missed triggers due to suspended systems are properly handled with Systemd timers: https://wiki.archlinux.org/title/Systemd/Timers#Realtime_timer

Maybe this edge case is worth implementing a Systemd timer.

File: modules/wlroots/hm.nix
Comment (trueNAHO) [reply modules/wlroots/hm.nix:29]: Maybe we should just keep `PATH` in `Environment`. Feel free to resolve this discussion.

File: lib/constructors.nix
Comment (danth) [reply lib/constructors.nix:22]: Done in c9a98dd2db53ec90943a1bc6232ab0cac5e59cfb

File: modules/rofi/hm.nix
Comment (trueNAHO) [reply modules/rofi/hm.nix:10]: > `with config.lib.stylix;` may be more appropriate

Do you mean this:

```suggestion
  finalString = with config.lib.stylix; ''
   * { background: rgba ( {{base00-rgb-r}}, {{base00-rgb-g}}, {{base00-rgb-b}}, ${popupsOpacity-int} % );
       lightbg: rgba ( {{base01-rgb-r}}, {{base01-rgb-g}}, {{base01-rgb-b}}, ${popupsOpacity-int} % );
       red: rgba ( {{base08-rgb-r}}, {{base08-rgb-g}}, {{base08-rgb-b}}, ${popupsOpacity-int} % );
       blue: rgba ( {{base0D-rgb-r}}, {{base0D-rgb-g}}, {{base0D-rgb-b}}, ${popupsOpacity-int} % );
       lightfg: rgba ( {{base06-rgb-r}}, {{base06-rgb-g}}, {{base06-rgb-b}}, ${popupsOpacity-int} % );
       foreground: rgba ( {{base05-rgb-r}}, {{base05-rgb-g}}, {{base05-rgb-b}}, ${popupsOpacity-int} % );
```

Note that I did not test this.

File: lib/types.nix
Comment (danth) [inline lib/types.nix:43]: This replaces the old `config.lib.stylix.scheme` which was effectively an alias to `config.lib.stylix.colors`. For backwards compatibility we could move all the types into `config.lib.stylix.types`, then it will be possible to bring back `scheme` as an alias to the new `config.stylix.colors` option, with a warning similar to what `config.lib.stylix.colors` does now.

Comment (danth) [reply lib/types.nix]: Fixed in 5d666dc885f430566f7204fcba1c5f39a9b97ac6

File: modules/wlroots/hm.nix
Comment (danth) [inline modules/wlroots/hm.nix:66]: This will fail for file names with spaces. Can probably be fixed using `lib.escapeShellArg`

File: modules/xorg/command.nix
Comment (danth) [inline modules/xorg/command.nix:14]: Same as for wlroots

File: lib/utils.nix
Comment (danth) [inline lib/utils.nix:1]: Converting these to functions (like `toHex`, `toInt`, `toPolarity`) will make our `lib` folder only depend on `pkgs` and the upstream `lib`, without reading anything from `config`. I'm hoping that will allow it to be passed as a module argument without causing infinite recursion.

File: lib/utils.nix
Comment (danth) [inline lib/utils.nix:41]: Duplicated function.

**Comment by Sntx626** [comment]
What's the state of this PR (draft)?

I'd be willing to help by rewriting and deduplicating the functions in `lib/utils.nix`.
Is there something else I can/should work on?

**Comment by danth** [comment]
@Sntx626 The main reason this isn't merged yet is because I'd like to remove the need to write out `config.lib.stylix` before every function, and just have it imported as `{ stylix, ... }`. I tried to implement that a while back, but it led to infinite recursion.

The things you mentioned should help towards that.

It would also be useful to have more testing to make sure this doesn't break existing configs, or at least gives a helpful error message.

**Comment by trueNAHO** [comment]
> The main reason this isn't merged yet is because I'd like to remove the need to write out `config.lib.stylix` before every function, and just have it imported as `{ stylix, ... }`. I tried to implement that a while back, but it led to infinite recursion.

Considering that several internal feature PRs are waiting for this PR to land, it might be better to refactor `config.lib.stylix` with `{ stylix, ... }` at a later time in a followup PR.

> It would also be useful to have more testing to make sure this doesn't break existing configs, or at least gives a helpful error message.

However, considering the enormous rewrite, this should definitely happen before merging this PR.

Personally, this PR and its direct dependants might be one of the most drastic improvements to this project so far. Thanks for all the work so far!

For reference, the following PRs and issues depend on this PR:

- https://github.com/danth/stylix/issues/131
- https://github.com/danth/stylix/issues/182
- https://github.com/danth/stylix/issues/208
- https://github.com/danth/stylix/issues/63
- https://github.com/danth/stylix/issues/63
- https://github.com/danth/stylix/pull/157

For reference, the following PRs and issues should be improved based on this PR:

- https://github.com/danth/stylix/pull/135
- https://github.com/danth/stylix/pull/175

Feel free to add any PRs or issues I missed and to update the description of this PR to properly link the PRs and issues.

File: lib/types.nix
Comment (trueNAHO) [inline lib/types.nix:1]: As mentioned in https://github.com/danth/stylix/issues/251#issuecomment-1932029384, the "black box" type introduced in https://github.com/danth/stylix/pull/102/commits/7a080cb74d3606299a637e7564fbb16e0de0a5a6 is a genius type conversion abstraction.

However, I have my performance doubts about providing conversion functions instead of converted values. Specifically, does calling a conversion function multiple times convert multiple times or only once? Would static attribute sets not provide vastly better performance by caching their result? Additionally, the immutably converted data prevents any kind of conversion inconsistencies.

File: lib/utils.nix
Comment (Sntx626) [inline lib/utils.nix:17]: Currently the function takes the `color` attr set.
This makes it relatively simple to call the function.

However this argument is not type checked. This will result in misleading errors at best and is documented nowhere atm. (except for this comment...)

Should we take `red`, `green` and `blue` as arguments instead?

File: lib/utils.nix
Comment (Sntx626) [inline lib/utils.nix:8]: Should we simplify this function?

I.e. `builtins.ceil (opacity * 255)`

File: lib/utils.nix
Comment (Sntx626) [inline lib/utils.nix:4]: We should probably create a library independent of config and tell the user how to extend their lib with ours.

This way we can finally be free of the config argument.

**Review by Sntx626 — COMMENTED** [review]
While functional, there's still room for improvement.

File: lib/types.nix
Comment (Sntx626) [reply lib/types.nix:1]: This is a valid concern.

However I think we should first measure the real-world impact of using the black box type and see if it is as bad as you think it could be. Or if there are other areas we should improve on first.

I also don't know enough about nix, but it could cache the result of the function and basically need to evaluate it only once. The optimization we seek might not even lay in this codebase.

Just an example of what would improve build times more:
I'm overriding all the colors myself. But stylix still generates a color palette from the bg image. I don't know if there's a way to turn that off, however running the genetic algorithm takes a lot of times. Which is not needed in my case afaik.

File: lib/utils.nix
Comment (danth) [reply lib/utils.nix:4]: A completely separate library (`stylix` / `stylixLib` or something) may be easier than trying to extend the existing `lib`.

File: lib/utils.nix
Comment (trueNAHO) [reply lib/utils.nix:8]:
```suggestion
      builtins.ceil (opacity * 255)
    )
```

File: .gitignore
Comment (trueNAHO) [inline .gitignore:2]: Early merging this file in https://github.com/danth/stylix/pull/291 benefits the project right away. The merge commit would remove this file:

```suggestion
```

**Review by trueNAHO — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (danth) [reply .gitignore:2]: #291 is now merged

File: lib/utils.nix
Comment (SomeGuyNamedMay) [reply lib/utils.nix:4]: I think a separate library is probably a better idea, extending lib doesnt seem to be worth the effort

**Pull request closed without merging**
